### ISSUE TITLE
ST train example doesn't work on multi gpu

### ISSUE THREAD

**Issue by turtle0x1** [AnswerDotAI/ModernBERT#147]
When trying to run https://github.com/AnswerDotAI/ModernBERT/blob/main/examples/train_st.py on a machine with multiple GPU it crashes with

`RuntimeError: Detected that you are using FX to symbolically trace a dynamo-optimized function. This is not supported at the moment.`

Its the same with most existing hugging face examples, will it / should multi GPU be supported?

Using the CUDA_VISIBLE_DEVICES is how I worked around it to train with one GPU.

**Comment by NohTow** [comment]
Hello,

To run multi-gpu training with ST, I am usually using `accelerate launch --num_processes num_gpu train_st.py`
You can also use `torchrun --nproc_per_node=num_gpu train_st.py`
Additional bonus, [it will use DDP instead of DP, which should be better.](https://sbert.net/docs/sentence_transformer/training/distributed.html)

I might want to debug at some point what is happening with plain python call, but if you are saying that it is happening with most models, then it is most likely an HF issue!

**Comment by turtle0x1** [comment]
Ah "secret" incantations, much appreciated that seems to have done it :)

Do you know where I should file a proper issue to support it using a plain python call like other models? :smile:

**Comment by NohTow** [comment]
Well, back in the days when the plain python call was working, there was a warning linking to the documentation, so it should not be this secret! (and I also use this in all the readme I write, the ones of this exact folder and the ones of PyLate, so trying to make it as known as possible)

About the issue, if it also happens with other models, I think this is more an transformer-related issue, so you might want to check there if there is already one opened on this subject and open one if not!
Otherwise, at some point I will try to have a look but not in the near future!

**Comment by turtle0x1** [comment]
Sorry for confusion, this doesn't happen with other models and by "secret" I meant "I hadn't read enough and missed the warning, forgive my ignorance".

Other training scripts with "old" bert models will use multiple GPU's with a standard "python" invocation, this doesn't work out the box but with your advise above I can use multiple GPU's :smile: .

**Comment by NohTow** [comment]
Yeah I was just realizing that I misunderstood your original message 
> Its the same with most existing hugging face examples
You meant the script is the same, not that the issue is the same, my bad.
Then we should investigate and fix this!
I don't think this is necessary to open another issue, the problem is pretty clear and I actually had the same error earlier, so we should be able to replicate!

**Comment by NohTow** [comment]
I just had a quick look at the issue.
Based on the error message it seems like this arise when the compilation is [called in different threads](https://github.com/pytorch/pytorch/issues/126024) and actually because the [`@torch.compile` decorator we are using to compile part of the model is not thread safe](https://github.com/pytorch/pytorch/issues/118387).

I did not go into the details of the calls with the different launcher but I suspect that the DP one (using `python`) call the compilation on the same model in the different thread while accelerate and torchrun don't.
 
So from my perspective we can advise people to use accelerate and torchrun until torch make `@torch.compile` thread safe. This seems ok to me, as in any case DDP should be preferred to DP. This might be a pain as practitioners will most likely just run the `python` call out-of-the-box and face the issue.
Maybe we could add some information in the boilerplates (as comment in the files), or add some information in the ST  warning (e.g, `Currently using DataParallel (DP) for multi-gpu training, while DistributedDataParallel (DDP) is recommended for faster training and can cause issue with compiled model`). Maybe another solution is to compile the part of the model in a thread safe way.

cc @tomaarsen @warner-benjamin

**Comment by shimengfeng** [comment]
@NohTow I want to check is the only solution is to run the code with accelerate at this point or are you plan to make any further adjustment in the code? I am encountering similar issue but rather than using ST but just using huggingface trainer.

**Comment by NohTow** [comment]
Well the issue seems more like a PyTorch issue than our code exactly (as there is an issue open to fix it).
We could maybe change how we compile the model to get rid of the decorators, but I don't think that is in our plan right @warner-benjamin?


If you are having the same issue, you should be able to run multi-gpu training using DP, or you'll have to use accelerate/torchrun to use DDP. Is there any issue with that?
